plugins/outputs/iotdb/iotdb.go

```diff
@@ -0,0 +1,276 @@
+		for _, field := range metric.FieldList() {
+			datatype, value := s.getDataTypeAndValue(field.Value)
+			if datatype == client.UNKNOW {
+				return nil, fmt.Errorf("datatype of %q is unknown, values: %v", field.Key, field.Value)
+			}
+			keys = append(keys, field.Key)
+			values = append(values, value)
+			dataTypes = append(dataTypes, datatype)
+		}
+		// Convert timestamp into specified unit
+		ts, err := s.convertTimestampOfMetric(metric)
+		if err != nil {
+			return nil, err
+		}
+		timestampList = append(timestampList, ts)
+		// append all metric data of this record to lists
+		deviceidList = append(deviceidList, metric.Name())
+		measurementsList = append(measurementsList, keys)
+		valuesList = append(valuesList, values)
+		dataTypesList = append(dataTypesList, dataTypes)
+		tagsList = append(tagsList, tags)
+	}
+	rwt := &recordsWithTags{
+		DeviceIDList:     deviceidList,
+		MeasurementsList: measurementsList,
+		ValuesList:       valuesList,
+		DataTypesList:    dataTypesList,
+		TimestampList:    timestampList,
+		TagsList:         tagsList,
+	}
+	return rwt, nil
+}
+
+// modify recordsWithTags according to 'TreatTagsAs' Configuration
+func (s *IoTDB) modifyRecordsWithTags(rwt *recordsWithTags) error {
+	switch s.TreatTagsAs {
+	case "fields":
+		// method 1: treat Tag(Key:Value) as measurement
+		for index, tags := range rwt.TagsList { // for each record
+			for _, tag := range tags { // for each tag of this record, append it's Key:Value to measurements
+				datatype, value := s.getDataTypeAndValue(tag.Value)
+				if datatype == client.UNKNOW {
+					return fmt.Errorf("datatype of %q is unknown, values: %v", tag.Key, value)
+				}
+				rwt.MeasurementsList[index] = append(rwt.MeasurementsList[index], tag.Key)
+				rwt.ValuesList[index] = append(rwt.ValuesList[index], value)
+				rwt.DataTypesList[index] = append(rwt.DataTypesList[index], datatype)
+			}
+		}
+		return nil
+	case "device_id":
+		// method 2: treat Tag(Key:Value) as subtree of device id
+		for index, tags := range rwt.TagsList { // for each record
+			subfix := ""
+			for _, tag := range tags { // for each tag, append it's Value
+				subfix = subfix + "." + tag.Value
+			}
+			rwt.DeviceIDList[index] = rwt.DeviceIDList[index] + subfix
+		}
+		return nil
+	default:
+		// something go wrong. This configuration should have been checked in func Init().
+		return fmt.Errorf("unknown 'convert_tags_to' method: %q", s.TreatTagsAs)
+	}
+}
+
+// Write records with tags to IoTDB server
+func (s *IoTDB) WriteRecordsWithTags(rwt *recordsWithTags) error {
+	// deal with tags
+	modifyError := s.modifyRecordsWithTags(rwt)
+	if modifyError != nil {
+		return modifyError
+	}
+	// write to IoTDB server
+	status, err := s.session.InsertRecords(rwt.DeviceIDList, rwt.MeasurementsList,
+		rwt.DataTypesList, rwt.ValuesList, rwt.TimestampList)
+	if status != nil {
+		if verifyResult := client.VerifySuccess(status); verifyResult != nil {
+			s.Log.Debug(verifyResult)
+		}
+	}
+	return err
+}
```

Comment:
This is still exported...
```suggestion
// Write records with tags to IoTDB server
func (s *IoTDB) writeRecordsWithTags(rwt *recordsWithTags) error {
	// deal with tags
	if err := s.modifyRecordsWithTags(rwt); err != nil {
		return err
	}
	// write to IoTDB server
	status, err := s.session.InsertRecords(rwt.DeviceIDList, rwt.MeasurementsList,
		rwt.DataTypesList, rwt.ValuesList, rwt.TimestampList)
	if status != nil {
		if verifyResult := client.VerifySuccess(status); verifyResult != nil {
			s.Log.Debug(verifyResult)
		}
	}
	return err
}
```

Reply:
it's fixed. Thanks a lot for reminding me.